control-plane/plugin/src/resources/snapshot.rs

```diff
@@ -0,0 +1,95 @@
+use crate::{
+    operations::GetSnapshots,
+    resources::{utils, SnapshotId, VolumeId},
+    rest_wrapper::RestClient,
+};
+use async_trait::async_trait;
+
+use crate::resources::utils::{CreateRow, GetHeaderRow};
+use prettytable::Row;
+
+/// A collection of VolumeSnapshot resource for all volumes.
+#[derive(clap::Args, Debug)]
+pub struct VolumesSnapshots {}
+
+/// A collection of VolumeSnapshot resource for a given volume.
+#[derive(clap::Args, Debug)]
+pub struct VolumeSnapshots {}
+
+#[derive(Debug, Clone, clap::Args)]
+/// BlockDevice args
+pub struct VolumeSnapshotArgs {
+    #[clap(long)]
+    /// Uuid of the volume (Optional).
+    volume: Option<VolumeId>,
+    #[clap(long)]
+    /// Uuid of the snapshot (Optional).
+    snapshot: Option<SnapshotId>,
+}
+
+impl VolumeSnapshotArgs {
+    /// get the volume id
+    pub fn volume(&self) -> Option<VolumeId> {
+        self.volume
+    }
+
+    /// get the snapshot id
+    pub fn snapshot(&self) -> Option<SnapshotId> {
+        self.snapshot
+    }
+}
+
+impl CreateRow for openapi::models::VolumeSnapshot {
+    fn row(&self) -> Row {
+        let state = &self.state;
+        row![
+            state.uuid,
+            state.creation_timestamp,
+            state.size,
+            state.source_volume
+        ]
+    }
+}
+
+#[async_trait(?Send)]
+impl GetSnapshots for VolumeSnapshots {
+    type VID = Option<VolumeId>;
+    type SID = Option<SnapshotId>;
+    async fn get_snapshots(volid: &Self::VID, snapid: &Self::SID, output: &utils::OutputFormat) {
+        if let Some(snapshots) = get_snapshots(volid, snapid).await {
+            // Print table, json or yaml based on output format.
+            utils::print_table(output, snapshots);
+        }
+    }
+}
+
+impl GetHeaderRow for openapi::models::VolumeSnapshot {
+    fn get_header_row(&self) -> Row {
+        (*utils::SNAPSHOT_HEADERS).clone()
+    }
+}
+
+async fn get_snapshots(
+    volid: &Option<VolumeId>,
+    snapid: &Option<SnapshotId>,
+) -> Option<Vec<openapi::models::VolumeSnapshot>> {
+    // XXX: 500 is a ballpark number. Once pagination is implemented, we'll reduce this
+    // to max entries per call.
+    let mut snapshots = Vec::with_capacity(500);
```

Comment:
Why not `Vec::new();`?

Reply:
As I interpret the Vec documentation - if we take a Vec::new() here and push each entry from `s.entries` that'll be slow as it'll have to reallocate each time to increase capacity, hence slower. That is avoided by preallocated capacity. Only if `s.entries` has more than this preallocated capacity, it'll go through reallocation.